### PR TITLE
Extract ProjectDescriptor from IProjectProvider

### PR DESCRIPTION
**Pull request by smoothdeveloper**
see #1440 #1411 

I used extension members for now to not have to change more code.

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
Trying to look at the 2 failing tests in `FindReferencesCommandTests` but can't figure out why it doesn't work, I'm able to use find references in VS fine.

Any idea?

File: src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs [hunk 1]
```diff
@@ -156,17 +156,18 @@ type internal ProjectProvider
...
-
+    let projectDescriptor = 
+        lazy
+            { IsForStandaloneScript = false
+              ProjectFile           = projectFileName.Value
+              TargetFramework       = targetFramework.Value
+              CompilerVersion       = None // This option is only relevant in scripts
+              CompilerOptions       = compilerOptions.Value
+              SourceFiles           = sourceFiles.Value
+              FullOutputFilePath    = fullOutputPath.Value
```
Comment (vasily-kirichenko) [inline src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs]: Why all the fields are lazy? IProjectProvider.Project is already lazy, so no point in more laziness here.

File: src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs [hunk 2]
```diff
@@ -189,15 +190,19 @@ type internal VirtualProjectProvider (source: string, filePath: string, vsVersio
...
+    let projectDescriptor = 
+        lazy
+            { IsForStandaloneScript = false
+              ProjectFile           = projectFileName
+              TargetFramework       = FSharpTargetFramework.NET_4_5
+              CompilerVersion       = Some compilerVersion
+              CompilerOptions       = flags
+              SourceFiles           = [| filePath |]
+              FullOutputFilePath    = Some (Path.ChangeExtension(projectFileName, ".dll"))
+            }
```
Comment (vasily-kirichenko) [inline src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs]: However, I don't see why it's lazy either.

File: src/FSharp.Editing/ProjectSystem/SolutionProvider.fs [hunk 3]
```diff
@@ -3,14 +3,30 @@
...
+    abstract Project: ProjectDescriptor
     abstract GetReferencedProjects: unit -> IProjectProvider list
     abstract GetAllReferencedProjectFileNames: unit -> string list 
     abstract GetProjectCheckerOptions: LanguageService -> Async<FSharpProjectOptions>
+
+[<AutoOpen>]
+module TypeExtensions =
+  // this is to avoid unneeded code churn while we figure out further changes to IProjectProvider
+  type IProjectProvider
+    with
```
Comment (vasily-kirichenko) [inline src/FSharp.Editing/ProjectSystem/SolutionProvider.fs]: Put `with` on the same line with `type IProjectProvider`

File: src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs [hunk 4]
```diff
@@ -189,15 +190,19 @@ type internal VirtualProjectProvider (source: string, filePath: string, vsVersio
         | _ -> FSharpCompilerVersion.FSharp_4_0
     let projectFileName = filePath + ".fsproj"
     let flags = [| "--noframework"; "--debug-"; "--optimize-"; "--tailcalls-" |]
-
+    
+    let projectDescriptor = 
+        lazy
+            { IsForStandaloneScript = false
```
Comment (smoothdeveloper) [inline src/FSharp.Editing.VisualStudio/ProjectSystem/ProjectProvider.fs]: bug seems to be here I switched to false while it was true.

File: src/FSharp.Editing/ProjectSystem/SolutionProvider.fs [hunk 5]
```diff
@@ -3,14 +3,30 @@
...
+      member x.IsForStandaloneScript = x.Project.IsForStandaloneScript
+      member x.ProjectFileName       = x.Project.ProjectFile
+      member x.TargetFramework       = x.Project.TargetFramework
+      member x.CompilerVersion       = x.Project.CompilerVersion
+      member x.CompilerOptions       = x.Project.CompilerOptions
+      member x.SourceFiles           = x.Project.SourceFiles
+      member x.FullOutputFilePath    = x.Project.FullOutputFilePath
```
Comment (vasily-kirichenko) [inline src/FSharp.Editing/ProjectSystem/SolutionProvider.fs]: Why the code is full with stuff like `project.Project.ProjectFile/etc.` having these properties existed?

**Pull request merged**